# My first Data Science project

**This project is designed to help a team of developers understand what apps are dominating the market and in which direction should they go**

**The goal of this project is to understand the importance of data analysis, how to analyse data efficiently and effectively.**

# Part 1. Data Exploring

This function helps us to explore data in different data sets, which makes it reusable

In [1]:
def explore_data(dataset, start, end, row_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') #adds a new line after each row
    if row_and_columns:
        print("Number of rows: ", len(dataset))
        print("Number of columns: ", len(dataset[0]))

In [2]:
opened_file = open("AppleStore.csv", encoding='utf8')
from csv import reader
reader_file = reader(opened_file)
app_store_list = list(reader_file)
app_store_header = app_store_list[0]
app_store = app_store_list[1:]

explore_data(app_store, 0, 10, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']


['5', '282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+', 'Reference', '37', '5', '45', '1']


['6', '283619399', 'Shanghai Mahjong', '10485713', 'USD', '0.99', '8253', '5516', '4', '4', '1.8', '4+', 'Games', '47', '5', '1', '1']


['7', '283646709', 'PayPal - Send and request money safely', '227795968', 'US

In [3]:
opened_file_1 = open("googleplaystore.csv", encoding='utf8')
reader_file_1 = reader(opened_file_1)
google_store_list = list(reader_file_1)
google_store_header = google_store_list[0]
google_store = google_store_list[1:]

explore_data(google_store, 0, 10, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+

**Let's have headers of both datasets in a seperated cell for visual representation**
> Table 1. Header row representation of AppleStore.csv (not full)

| index_number | Label            | Description                       |
|--------------|------------------|-----------------------------------|
| 1            | id               | Id of the app in AppleStore       |
| 2            | track_name       | Name of the app                   |
| 3            | size_bytes       | Size of the app                   |
| 4            | currency         | Price currency of the app         |
| 5            | price            | Price of the app                  |
| 6            | rating_count_tot | Total rating of the app           |
| 7            | rating_count_ver | Rating of the app by version      |
| 8            | user_rating      | User rating of the app            |
| 9            | user_rating_ver  | User rating of the app by version |
| 10           | ver              | Current version of the app        |
| 12           | cont_rating      | Content                               |
| 12           | prime_genre      | Genre of the app                  |

> Table 2. Header row representation of googleplaystore.csv (not full)

| index_number | Label          |
|--------------|----------------|
| 0            | App            |
| 1            | Category       |
| 2            | Rating         |
| 3            | Reviews        |
| 4            | Size           |
| 5            | Installs       |
| 6            | Type           |
| 7            | Price          |
| 8            | Content Rating |
| 9            | Genres         |
| 10           | Last Updated   |
| 11           | Current Ver    |
| 12           | Android Ver    |


In [6]:
print(google_store[10472])
print(google_store[0])
print('\n')
print(google_store_header)


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [5]:
del google_store[10472]

In [10]:
duplicate_apps = []
unique_apps = []
for app in google_store:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print("Number of duplicate apps: ", len(duplicate_apps))
print("Number of unique apps: ", len(unique_apps))

Number of duplicate apps:  1181
Number of unique apps:  9659


# Part 2. Data cleaning

## Cleaning google_store dataset
**Let's start by creating a dictionary**

In [12]:
#create a dictionary for each duplicate app with name-review (key-value) pair
reviews_max = {}
for app in google_store:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


With the help of a dictionary we could find apps with the heighest number of reviews (among the same name apps). The heighest number of reviews shows us an original app.

In [19]:
android_clean = []
already_added = []
for app in google_store:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
explore_data(android_clean, 0, 10, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+

Therefore, we could use it to clean our initial dataset (google_store) and make a new clean android_clean dataset by comparing dictionary key-value pairs with the initial dataset.

## Remove non-English apps
**To make our data clean, we need to remove non-English apps (apps that have more than 3 non-english characters)**
1. We define a function. It checks an app name and returns true if the app name contains less than 3 non-english characters
2. We clean our data by employing the function and save it

In [22]:
def check_english(string):
    n_ascii = 0
    for letter in string:
        if ord(letter) > 127:
            n_ascii += 1
    if n_ascii > 3:
        return False
    return True

False


In [26]:
ios_english = []
android_english = []
for app in app_store:
    name = app[2]
    if check_english(name):
        ios_english.append(app)
for app in android_clean:
    name = app[0]
    if check_english(name):
        android_english.append(app)
print(len(ios_english))
print(len(android_english))

6183
9614


## Isolating non-free apps

In [34]:
ios_english_free = []
android_english_free = []
for app in ios_english:
    price = float(app[5])
    if price == 0.0:
        ios_english_free.append(app)
for app in android_english:
    price = app[7]
    if price == "0":
        android_english_free.append(app)
print(len(ios_english_free))
print(len(android_english_free))

3222
8864


Isolating the free apps will be our last step in the data cleaning process. On the next cells, we're going to start analyzing the data.

# Part 3. Data Analysis

**In this part, we are going to analyse our data based on different parameters i.e. genres, content ratings etc.**

In [35]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for each in dataset:
        total += 1
        value = each[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    result = {}
    for key in table:
        value = (table[key]/total)*100
        result[key] = value
    return result

In [36]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [43]:
# display the frequency table of genres
prime_genre_ios = display_table(ios_english_free, 12)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.31678700

- It is evident from the results above that the dominating genre on iOS is Games (58% of total)
- The second most common genre is Entertaiment with almost 8% of total genres
- Surprisingly, Social Networking takes only 3.3% of total genres and gives in it's position to Photo & Video AND Education

- **Generally speaking, we can see that the app designed for practical purposes (education, shopping, utilities, productivity, lifestyle) are much less used as compared to the entertainment apps (games, photo and video, social networking, sports, music). It is due to the modern society priorities and lifestyle.**
- **However, I cannot recommend any app based only on this frequency table. But certainly, I can say that Games genre has the highest number of people using/playing it.&**

In [ ]:
genres_android = display_table(android_english_free, 9)

According to these results, 
- the most frequent genre in GooglePlayStore is Tools (which takes about 8.45% of all genres)
- suprisingly, Education comes alongside with Entertaiment in the short range (about 0.7% difference)

Moreover, comparing to the results from AppStore, it is clearly observable that Entertainment in GooglePlayStore is much less than in AppStore (about 2% less), but Education is much bigger than in AppStore (about 1.7% higher). It clearly indicates the balanced landscape of both practical and fun apps in GooglePlayStore

In [ ]:
# display the frequency table of category
category_android = display_table(android_english_free, 1)

According to this list,
- Family category prevails over the runner-up (Game), taking twice than it
- Tools and Game genres come alongside with the minor difference between it (1.3%)